### ISSUE TITLE
Add addPlugin method to CLIEngine

### ISSUE THREAD

**Issue by wooorm** [eslint/eslint#1971]
Is it intentional that all plug-ins have a name which starts with `PLUGIN_NAME_PREFIX`? Or should it be possible to use plug-ins with other names too?

Currently, [`lib/cli-engine.js`](https://github.com/eslint/eslint/blob/6e976758533d73957fb5ceb65793f65d847fb28e/lib/cli-engine.js#L103) requires the prefix to be present, but to my knowledge the [docs](http://eslint.org/docs/developer-guide/working-with-plugins.html#share-plugins) do not specify this.

**Comment by lo1tuma** [comment]
Yes, this is intentional and specified in the first sentence of the [documentation](http://eslint.org/docs/developer-guide/working-with-plugins.html#working-with-plugins).

The reasons for this have been described in the original issue #1106:

> Each plugin is an npm module with a name in the format of eslint-plugin-<plugin-name>, such as eslint-plugin-jquery.
* This ensures uniqueness for all plugins published on the npm registry.
* It also allows users to easily search for and find plugins.
* The naming patterns also leaves the door open for creating other tools to easily find plugins.

Is there any use-case where you can't use the prefix?

**Comment by wooorm** [comment]
Oh right, sorry I totally missed that. I have a plug-in, `eslint-md`, which was created before #1832 (see #1817), and now I was hoping to update it without having to change names. I’ll accept that I’ll have to though, so you can close this. Thanks!

**Comment by wooorm** [comment]
Hmm I have to say that testing plugins is hard when you cannot specify the current project as a plug-in!

I first symlinked the project into its own `node_modules`, but coverage (istanbul) doesn’t like that obviously.

I also checked out `eslint-tester`, but that only exposes utilities for rules. Maybe this has more to do with the new processor plugins… I’ll work on this some more some other day, and let you know how things go.

**Comment by ilyavolodin** [comment]
@wooorm Check out some of the existing plugins if you have issues with testing (you can take a look at my plugin: eslint-plugin-backbone, which has 100% coverage).
Instead of symlinking project into eslint, your plugin should use `peerDependency` for eslint. `eslint-tester` is just a wrapper to make unit tests for rules DRY, under the hood it just uses mocha.

**Comment by wooorm** [comment]
Thanks for responding @ilyavolodin!

Re your points: I have checked other plugins to see how they test their coverage (with eslint-tester), however, as it just exposes ways to test rules, it’s not an applicable library for testing processing plug-ins. In fact, `eslint-tester` does not test the plug-in, and how the plug-in works with the CLI—rather, it tests individual rules (Right? I’m sure you guys know more about it than I do though!)

Also, with regards to the symlinking, I know about peer-dependencies, that’s all set up properly. However peer-dependencies have no effect when running `npm install` inside a project (it does when invoking `npm install eslint-plugin-some-name` though, of course), other than installing `eslint` inside the project’s `./node_modules` just like a dev-dependency would. This makes `eslint` require-able in the tests, but because `eslint` is inside the project, and `eslint-plugin-some-name` is not (rather at `../../` relative to ESLint), `eslint` cannot require it. And that’s why I had to use symlinking.

For an example on why this does not work, see ESLint’s processor tests, which use require stubs to work around this: [tests/lib/cli-engine.js#L37](https://github.com/eslint/eslint/blob/de62efa5de88da24378ec9942421f0b1078b6ba8/tests/lib/cli-engine.js#L37).

It’s a bit hard to describe the problem, I hope the above helped explain it though :smile:

**Comment by ilyavolodin** [comment]
@wooorm `eslint-tester` was created specifically to DRY rules testing. We discussed possibility of creating a wrapper for testing other thing, but it was deemed unnecessary. Here's an example of unit test for preprocessor (I haven't verify it, but it should point you in the right direction):

```js
var CLIEngine = require("eslint").CLIEngine;
var processor = require("path_to_your_processor_module");

describe('preprocessor test', function() {
    it('should be able to handle .md file', function() {
        var config = { processors: {
            '.md': {
                preprocess: processor.preprocess,
                postprocess: processor.postprocess
            }
        };
        var cli = new CLIEngine(config);

        var filename = 'path_to_md_file';

        var report = cli.executeOnFiles([filename]);

        assert.equal(report.results.errorCount, 1);
    });
});
```

**Comment by nzakas** [comment]
`CLIEngine` doesn't except `config` as an argument. It only supports command-line flags. See: http://eslint.org/docs/developer-guide/nodejs-api.html

You'd need to create a config file outside of the test.

**Comment by ilyavolodin** [comment]
Ahh... you are correct. Well, in that case, I don't think there's any other way to test custom processors other then using `proxyquire`. Since CLIEngine will user `require` to load all plugins. Sorry for confusion @wooorm

**Comment by nzakas** [comment]
We could add a method to load a plugin.

**Comment by ilyavolodin** [comment]
Yes, I think it might be necessary to expose `loadPlugins` on the `CLIEngine` level, otherwise it's going to be pretty tough to test processors.

**Comment by nzakas** [comment]
It would have to be `addPlugin(name, pluginobject)` since there is no config file to read from.

**Comment by smussell** [comment]
I agree I'd like to see something like this added.  My use case is slightly different.  I have a central project with my Gulpfile, which has an eslint task and specifies my plugins/rules.  I use npm to require that project in all other projects.  My gulp project currently has a peer dependency for "eslint-plugin-react" which makes everything work for now, but I do get this warning:

```
npm WARN peerDependencies The peer dependency eslint-plugin-react@^2.2.0 included
npm WARN peerDependencies longer be automatically installed to fulfill the peerDependency 
npm WARN peerDependencies in npm 3+. Your application will need to depend on it explicitly.
```
Which means eventually it will stop working and instead of updating all projects which depend on my gulp project it would be much easier to update my plugin config to point to its local plugin, then I can move the plugin from a peer dependency to a standard dependency.

**Comment by nzakas** [comment]
This issue is marked as accepted, which means it's waiting for implementation. It isn't a debate, we are doing it, we just need someone to make the change.
